### DISCUSSION TITLE
VirtualDOM이 필요한 이유에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by byhhh2** [woowacourse-study/2022-woowahan-bansanghwe#68] [React]
> 준비를 많이 못해서 쉬운 내용으로 해볼게요 !!!!

**Comment by byhhh2** [comment]
## 브라우저 렌더링 과정

1. 브라우저가 전달받은 HTML 파일을 브라우저 엔진이 이를 파싱하고 DOM node로 이루어진 트리를 생성한다.
2. CSS 파일과 각 element의 inline 스타일을 파싱하고 스타일 정보를 추가하여 render tree를 생성한다. 
3. render tree가 만들어지면 Layout 과정을 거친다. 
    - layout 과정에서 각 노드들의 스크린 좌표가 주어지고, 어디에 위치해야 할지가 정해진다.
4. 마지막으로 painting 과정이 실행된다. 앞선 과정에서 얻을 수 있는 정보들로 모든 요소들에 색을 입히는 과정이다. 
5. 위의 과정을 거치고 나면 화면에 UI가 렌더된다. 

## DOM 조작의 비효율성

DOM을 직접 조작하여 화면을 업데이트하면, 브라우저 렌더링 과정을 전부 진행하게 된다. 

DOM을 조작할 때마다 이런 과정이 반복되는 것은 실제로 많은 연산을 수반하고 비용이 많이 들기 때문에 프로그램의 성능을 저하시킨다. 

## SPA의 등장

예전에는 서버에서 데이터와 함께 완전한 페이지를 렌더해주는 SSR 방식을 많이 사용했다. 원래 DOM은 정적인 페이지를 보여주는 데에 사용됐기 때문에 큰 문제가 되지 않았다.

하지만 SPA의 등장과 함께 CSR 방식이 많이 사용되면서 DOM 업데이트가 상당히 많이, 복잡하게 발생하는 어플리케이션이 많아졌다. 

이런 맥락에서 DOM 조적을 통해 화면을 렌더링하는 과정의 비효율성을 해결하고 최적화를 할 필요성이 생겼다.

이런 과정에서 등장한 것이 **Virtual DOM**이다. 

## Virtual DOM이란?

Virtual DOM은 실제 DOM tree의 가벼운 버전이다. 

Virtual DOM은 실제 DOM tree를 복제한 자바스크립트 객체이다. 

가벼운 버전인 이유는 

- Virtual DOM은 실제 DOM과 같은 class, style 등의 속성들을 가지고 있지만, 화면에 변화를 직접줄 수 있는 getElementById과 같은 DOM API는 갖고 있지 않기 때문

## Virtual DOM의 동작 방식

브라우저에 실제 DOM tree가 생성되고 UI가 렌더됐을 때

Virtual DOM은 실제 DOM을 가벼운 버전으로 복사한다. 

그리고 DOM에 변화가 생기면 Virtual DOM은 다시 새로운 가상의 DOM tree를 처음부터 다시 만든다. 

## 새로운 가상 DOM tree를 만드는 것이 더 비효율적아닌가?

DOM node를 조작하는 것의 비효율성은 DOM 트리를 업데이트하는 과정에서 발생하는 것이 아닌, 렌더링하는 과정에서 비싼 비용이 드는 것이다.

하지만 **Virtual DOM은 렌더링을 하지 않고 메모리 상에서 트리를 변경하는 일**이기 때문에 빠르게 작업이 진행된다. 

Virtual DOM의 내부 구현체를 살펴보면,

**diff 함수**에서 이전 상태의 DOM 트리와 새로운 상태의 DOM 트리를 받아오고, 

이를 비교해 변경된 부분만을 확인한다. 

**patch 함수**에서 diff 함수를 통해 얻은 변경된 부분만 담겨있는 patches를 인자로 받는다. **이 변경된 사항만을 실제 DOM 노드에 적용하여 렌더링 과정을 수행한다.** 

이러한 과정을 봤을 때 Virtual DOM은 사실상 버퍼링 또는 캐싱의 역할을 한다. 

DOM을 조작할 때마다 렌더링 과정을 반복하는 것이 아니라 변화들을 전부 Virtual DOM에 반영한 후 변경된 부분만 모아서 실제 DOM에 적용되어 한번만 렌더링 하도록 하여 성능을 최적화한다. 

## React에서의 Virtual DOM

react를 통해 코드를 작성할 때 JSX 문법을 많이 사용하는데 

JSX는 바벨을 통해 자바스크립트로 변환되고 createElement 함수를 호출한다. 

이 함수를 통해 객체로 변환되고 이 객체를 활용해 Virtual DOM 트리를 구성하게 됩니다. 이 객체를 이용해 render() 함수를 호출하면 비로소 실제 DOM 요소가 된다. 

## 재조정 (reconciliation)

재조정이란 Virtual DOM과 실제 DOM을 비교하여 일치시키는 과정을 말한다. 

리액트는 변경 전 Virtual DOM과 변경 이후의 Virtual DOM을 모두 유지하고 있고, 이 두 트리의 스냅샷을 비교해서 변화된 부분만 실제 DOM에 적용한다.

## 참고

- [돔하디의 Virtual DOM](https://www.youtube.com/watch?v=6rDBqVHSbgM&t=320s)

**Comment by greenblues1190** [comment]
왜 DOM 요소가 변경될 때 바뀐 요소만 변경하지 않고 Virtual DOM 트리를 새로 구성할까요?

**Comment by greenblues1190** [comment]
React와 ReactDOM의 관계에 대해 설명해주세요.

**Comment by greenblues1190** [comment]
재조정 알고리즘은 React와 ReactDOM 중 어디에 있을까요?

**Comment by greenblues1190** [comment]
React는 왜 이름이 React일까요?

**Comment by airman5573** [comment]
VirtualDOM은 어떻게 만들어 지나요?

**Comment by airman5573** [comment]
setState는 왜 비동기적으로 작동할까요?

**Comment by airman5573** [comment]
VirtualDOM을 수정하고 DOM이랑 비교해서 다른점을 그린다고 하셨는데,
그러면 그냥 DOM을 수정하는게 더 빠르지 않나요?
